ton/src/emulators/thread_pool.rs

```diff
@@ -0,0 +1,348 @@
+    _phantom: std::marker::PhantomData<Obj>,
+}
+
+impl<Obj, Task, Retval> ThreadPool<Obj, Task, Retval>
+where
+    Obj: PooledObject<Task, Retval> + Send + 'static,
+    Task: Send + 'static,
+    Retval: Send + 'static,
+{
+    pub fn new(mut obj_arr: Vec<Obj>, cfg: ThreadPoolConfig) -> TonResult<Self> {
+        if obj_arr.is_empty() {
+            bail_ton!("Object array for ThreadPool is empty");
+        }
+
+        let num_threads = obj_arr.len();
+        let mut senders = Vec::with_capacity(num_threads);
+        let mut thread_handles = Vec::with_capacity(num_threads);
+        let mut cnt_jobs_in_queue = Vec::with_capacity(num_threads);
+        let mut cnt_done_tasks = Vec::with_capacity(num_threads);
+        let mut cnt_errored_tasks = Vec::with_capacity(num_threads);
+
+        for _ in 0..num_threads {
+            let (tx, rx): CommandChannel<Task, TonResult<Retval>> = mpsc::channel();
+            let obj = obj_arr.pop().unwrap();
+
+            let handle = thread::spawn(move || {
+                if let Some(init_fn) = cfg.pin_to_core_function {
+                    init_fn();
+                }
+                Self::worker_loop(obj, rx)
+            });
+            senders.push(tx);
+            thread_handles.push(handle);
+            cnt_jobs_in_queue.push(AtomicUsize::new(0));
+            cnt_done_tasks.push(AtomicUsize::new(0));
+            cnt_errored_tasks.push(AtomicUsize::new(0));
+        }
+        Ok(Self {
+            senders,
+            thread_handles,
+            cnt_jobs_in_queue,
+            cnt_done_tasks,
+            cnt_errored_tasks,
+            cnt_current_jobs: AtomicUsize::new(0),
+            default_timeout_emulation: cfg.default_timeout_emulation,
+            thread_queue_capacity: cfg.thread_queue_capacity,
+            _phantom: std::marker::PhantomData,
+        })
+    }
+    async fn increment_id_and_get_index(&self, deadline: u128) -> TonResult<usize> {
+        loop {
+            if get_now_ms() > deadline {
+                return Err(TonError::EmulatorPoolTimeout {
+                    deadline_time: deadline,
+                });
+            }
+            //  find thread with minimum queue size
+            let mut min_queue_value = self.thread_queue_capacity as usize + 1;
+            let mut target_queue_index = self.senders.len(); // set bad index
+            for i in 0..self.senders.len() {
+                let current_queue_size = self.cnt_jobs_in_queue[i].load(Ordering::Relaxed);
+                if current_queue_size < 2 {
+                    target_queue_index = i;
+                    break;
+                }
+                // as late as possible
+                if current_queue_size <= min_queue_value {
+                    min_queue_value = current_queue_size;
+                    target_queue_index = i;
+                }
+            }
+
+            if target_queue_index == self.senders.len() {
+                sleep(Duration::from_millis(1)).await;
```

Comment:
I believe 1ms is too short for sleep, I would use 2 or 5 (considering avg emulation time ~2ms)